### PR TITLE
feat(server): add user impl

### PR DESCRIPTION
**Pull request by fenngZ**
feat(server): add user impl

### CODE REVIEW

File: server/src/main/java/io/openbytes/odi/infrastructrue/repository/user/UserRepositoryImpl.java [hunk 1]
```diff
@@ -16,5 +16,65 @@
...
+
+
+    /**
+     * @param user
+     * @return
+     */
+    public UserPO toPO(User user) {
+        int status = 0;
+        if (user.getStatus() != null) {
+            status = user.getStatus().getCode();
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/infrastructrue/repository/user/UserRepositoryImpl.java:67]: null will be 0

File: server/src/main/java/io/openbytes/odi/domain/user/User.java [hunk 2]
```diff
@@ -73,7 +69,7 @@
             this.code = code;
         }
 
-        public static Optional<Status> get(int code) {
+        public static Optional<Status> Status(int code) {
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/domain/user/User.java:72]: of

File: server/src/main/java/io/openbytes/odi/application/GithubService.java [hunk 3]
```diff
@@ -43,6 +43,8 @@
 
     private static final String OAUTH_ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token";
 
+    private static final String GITHUB_USER = "https://api.github.com/user";
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/application/GithubService.java:46]: rename

File: server/src/main/java/io/openbytes/odi/application/GithubService.java [hunk 4]
```diff
@@ -43,6 +43,8 @@
...
+
     private static final MediaType MediaType_JSON = MediaType.get("application/json; charset=utf-8");
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/application/GithubService.java:49]: all capital

File: server/src/main/java/io/openbytes/odi/application/GithubService.java [hunk 5]
```diff
@@ -133,19 +134,53 @@ public GithubService(@Value("${github.account.clientId}") @NotNull String client
...
+     * if oauthToken get error or code is not 200, return a exception
+     * else, return decoded user
+     */
+    public GithubUser getUserByOauthToken(String oauthToken) {
+        if (StrUtil.isEmpty(oauthToken)) {
+            throw new BizException(CodeAndMessage.argsIllegal, "OAuthToken is empty");
+        }
+
+        Request request = new Request.Builder().header("Authorization", "token " + oauthToken).url(GITHUB_USER).get().build();
+        // 2. get response
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/application/GithubService.java:169]: improve notes, or consider remove

File: server/src/main/java/io/openbytes/odi/application/UserService.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        //get github user
+        //will return exception if githubUser is not correct
+        GithubUser githubUser = githubService.getUserByOauthToken(oauthAccessToken.getAccessToken());
+
+        // check user exist?
+
+        Optional<ChannelUserPO> channelUserPO = userRepository.getUserByChannelTypeAndChannelUserId("github", githubUser.getId());
+        if (!channelUserPO.isEmpty() && StrUtil.isNotEmpty(channelUserPO.get().getUserId())) { //非空，则返回
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/application/UserService.java:59]: !channelUserPO.isEmpty -> channelUserPO.isPresent()

File: server/src/main/java/io/openbytes/odi/application/UserService.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            Optional<User> odiUser = userRepository.getById(channelUserPO.get().getUserId());
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/application/UserService.java:60]: maybe you can:
Optional<User> odiUser = userRepository.getUserByChannelTypeAndChannelUserId("github", githubUser.getId())
.filter(l -> StrUtil.isNotEmpty(l.getUserId()))
.map(l -> userRepository.getById(l.getUserId()))
if (odiUser.isPresent) {
    return new BindByGithubDeviceCodeResult(BindByGithubDeviceCodeResult.Status.SUCCESS, odiUser.get());
}

**Review by ziyiyizi — CHANGES_REQUESTED** [review]

File: server/src/main/java/io/openbytes/odi/application/GithubService.java (hunk 4 above)
Comment (fenngZ) [reply server/src/main/java/io/openbytes/odi/application/GithubService.java:49]: done

File: server/src/main/java/io/openbytes/odi/domain/user/GithubOauthTokenResult.java [hunk 8]
```diff
@@ -60,6 +72,15 @@
...
         @JSONField(name = "error_uri")
         private final String errorUri;
+
+        public static OauthFailed of(String error) {
+            for (OauthFailed oauthFailed : OauthFailed.values()) {
+                if (oauthFailed.getError().equals(error)) {
+                    return oauthFailed;
+                }
+            }
+            return null;
```
Comment (ziyiyizi) [inline server/src/main/java/io/openbytes/odi/domain/user/GithubOauthTokenResult.java:82]: not recommend return null, maybe improve in future

**Review by ziyiyizi — APPROVED** [review]
1

**Pull request merged**